cookiejar.go

```diff
@@ -0,0 +1,221 @@
+package fasthttp
+
+import (
+	"bytes"
+	"net"
+	"sync"
+	"time"
+)
+
+// CookieJar manages cookie storage
+type CookieJar struct {
+	m           sync.Mutex
+	hostCookies map[string][]*Cookie
+}
+
+func (cj *CookieJar) init() {
+	cj.m.Lock()
+	{
+		if cj.hostCookies == nil {
+			cj.hostCookies = make(map[string][]*Cookie)
+		}
+	}
+	cj.m.Unlock()
+}
+
+func copyCookies(cookies []*Cookie) (cs []*Cookie) {
+	// TODO: Try to delete the allocations
+	cs = make([]*Cookie, 0, len(cookies))
+	for _, cookie := range cookies {
+		c := AcquireCookie()
+		c.CopyTo(cookie)
+		cs = append(cs, c)
+	}
+	return
+}
+
+// Get returns the cookies stored from a specific domain.
+//
+// If there were no cookies related with host returned slice will be nil.
+//
+// returned cookies can be released safely.
+func (cj *CookieJar) Get(uri *URI) (cookies []*Cookie) {
+	if uri != nil {
+		cookies = cj.get(uri.Host(), uri.Path())
+	}
+	return
+}
+
+func (cj *CookieJar) get(host, path []byte) (rcs []*Cookie) {
+	cj.init()
+
+	var (
+		err     error
+		cookies []*Cookie
+		hostStr = b2s(host)
+	)
+	// port must not be included.
+	hostStr, _, err = net.SplitHostPort(hostStr)
+	if err != nil {
+		hostStr = b2s(host)
+	}
+	cj.m.Lock()
+	{
+		// get cookies deleting expired ones
+		cookies = cj.getCookies(hostStr)
+	}
+	cj.m.Unlock()
+
+	rcs = make([]*Cookie, 0, len(cookies))
+	for i := 0; i < len(cookies); i++ {
+		cookie := cookies[i]
+		if len(path) > 1 && len(cookie.path) > 1 && !bytes.HasPrefix(cookie.Path(), path) {
+			continue
+		}
+		rcs = append(rcs, cookie)
+	}
+
+	return
+}
+
+// getCookies returns a cookie slice releasing expired cookies
+func (cj *CookieJar) getCookies(hostStr string) []*Cookie {
+	var (
+		cookies = cj.hostCookies[hostStr]
+		t       = time.Now()
+		n       = len(cookies)
+	)
+	for i := 0; i < len(cookies); i++ {
+		c := cookies[i]
+		if !c.Expire().Equal(CookieExpireUnlimited) && c.Expire().Before(t) { // cookie expired
+			cookies = append(cookies[:i], cookies[i+1:]...)
+			ReleaseCookie(c)
+			i--
+		}
+	}
+	// has any cookie been deleted?
+	if n > len(cookies) {
+		cj.hostCookies[hostStr] = cookies
+	}
+	return cookies
+}
+
+// Set sets cookies for a specific host.
+//
+// The host is get from uri.Host().
+//
+// If the cookie key already exists it will be replaced by the new cookie value.
```

Comment:
You should add a comment that makes clear the CookieJar will make a copy of the cookie to keep and it's up to the caller to release it's own cookie.